### DISCUSSION TITLE
ms to msprime conversion error (using demes.from_ms)

### DISCUSSION THREAD

**Discussion by mshpak76** [tskit-dev/msprime#2069] [Q&A] [answered]
I have some ms scripts that I'm trying to convert to msprime using the from_ms() function in the demes package. I tried testing it on a simple example of toy code and keep getting the same error, even though I essentially followed the same format as in the example given here:

https://github.com/tskit-dev/msprime/discussions/1897

I have:
    
    cmd = ("-I 3 5 5 5 -ej 0.25 3 2 -ej 0.75 2 1 -n 2 0.5 -n 3 0.2")
    demes.from_ms(cmd, N0=10000, deme_names=None)

which gives me the error message:

    TypeError: from_ms() takes 1 positional argument but 2 positional arguments (and 1 keyword-only argument)

There's evidently something subtly wrong with my syntax, though as far as I can tell my input argument format matches the example given in its format.

I have an additional question about from_ms() - rather than returning a graph, is there some way to get actual msprime code as output?

Thanks, Max

**Comment by grahamgower** [comment, marked as answer]
Hi @mshpak76. The error is because you need to specify the second parameter (and third) by name. So it needs to be `from_ms(cmd, N0=10000)` rather than `from_ms(cmd, 10000)`. Your edited code example works for me!

**Comment by mshpak76** [comment]
Thanks, it seems to work now (at least it doesn't throw an error).
As a follow-up, is there some way to see the msprime code that is generated
from ms in the conversion process rather than just getting the graph object
as output?

On Mon, Jun 13, 2022 at 9:38 AM Graham Gower ***@***.***>
wrote:

> Hi @mshpak76 <https://github.com/mshpak76>. The error is because you need
> to specify the second parameter (and third) by name. So it needs to be from_ms(cmd,
> N0=10000) rather than from_ms(cmd, 10000). Your edited code example works
> for me!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/tskit-dev/msprime/discussions/2069#discussioncomment-2939155>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AUZMZCOGFFPAH6Q3B2OEH6LVO5BURANCNFSM5YUM5LMA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
=======================
Max Shpak, Ph.D.
Department of Genetics
University of Wisconsin
Madison, WI 53706

**Reply by grahamgower** [reply]
The demes command line interface might be simpler in this case?
```
$ demes ms -N0 10000 -I 3 5 5 5 -ej 0.25 3 2 -ej 0.75 2 1 -n 2 0.5 -n 3 0.2 > model.yaml
$ cat model.yaml
time_units: generations
generation_time: 1
demes:
- name: deme1
  epochs:
  - {end_time: 0, start_size: 10000.0}
- name: deme2
  start_time: 30000.0
  ancestors: [deme1]
  epochs:
  - {end_time: 0, start_size: 5000.0}
- name: deme3
  start_time: 10000.0
  ancestors: [deme2]
  epochs:
  - {end_time: 0, start_size: 2000.0}
```

You can pass that file directly to msprime's command line interface
```
msp ancestry --demography model.yaml ...
```

And you can plot the yaml demography using [`demesdraw`](https://github.com/grahamgower/demesdraw).
```
demesdraw tubes model.yaml model.svg
```
![model](https://user-images.githubusercontent.com/8665077/173380907-2744142b-4580-4b32-8fcf-b46175fc20b2.svg)


If you really want/need to work in Python, then you can look at the `msprime.Demography` object constructed via from_demes(). There's no msprime code generated here, _per se_. But the Demography object can be directly inspected.

```python
import demes
import msprime

cmd = "-I 3 5 5 5 -ej 0.25 3 2 -ej 0.75 2 1 -n 2 0.5 -n 3 0.2"
graph = demes.from_ms(cmd, N0=10000)
demog = msprime.Demography.from_demes(graph)
print(demog.debug())
```

```
$ python /tmp/a.py
DemographyDebugger
╠══════════════════════════════════╗
║ Epoch[0]: [0, 1e+04) generations ║
╠══════════════════════════════════╝
╟    Populations (total=3 active=3)
║    ┌───────────────────────────────────────────────────────────────────┐
║    │       │     start│       end│growth_rate  │ deme1 │ deme2 │ deme3 │
║    ├───────────────────────────────────────────────────────────────────┤
║    │  deme1│   10000.0│   10000.0│ 0           │   0   │   0   │   0   │
║    │  deme2│    5000.0│    5000.0│ 0           │   0   │   0   │   0   │
║    │  deme3│    2000.0│    2000.0│ 0           │   0   │   0   │   0   │
║    └───────────────────────────────────────────────────────────────────┘
╟    Events @ generation 1e+04
║    ┌───────────────────────────────────────────────────────────────────────────────┐
║    │   time│type        │parameters        │effect                                 │
║    ├───────────────────────────────────────────────────────────────────────────────┤
║    │  1e+04│Population  │derived=[deme3],  │Moves all lineages from the 'deme3'    │
║    │       │Split       │ancestral=deme2   │derived population to the ancestral    │
║    │       │            │                  │'deme2' population. Also set 'deme3'   │
║    │       │            │                  │to inactive, and all migration rates   │
║    │       │            │                  │to and from the derived population to  │
║    │       │            │                  │zero.                                  │
║    └───────────────────────────────────────────────────────────────────────────────┘
╠══════════════════════════════════════╗
║ Epoch[1]: [1e+04, 3e+04) generations ║
╠══════════════════════════════════════╝
╟    Populations (total=3 active=2)
║    ┌───────────────────────────────────────────────────────────┐
║    │       │     start│       end│growth_rate  │ deme1 │ deme2 │
║    ├───────────────────────────────────────────────────────────┤
║    │  deme1│   10000.0│   10000.0│ 0           │   0   │   0   │
║    │  deme2│    5000.0│    5000.0│ 0           │   0   │   0   │
║    └───────────────────────────────────────────────────────────┘
╟    Events @ generation 3e+04
║    ┌───────────────────────────────────────────────────────────────────────────────┐
║    │   time│type        │parameters        │effect                                 │
║    ├───────────────────────────────────────────────────────────────────────────────┤
║    │  3e+04│Population  │derived=[deme2],  │Moves all lineages from the 'deme2'    │
║    │       │Split       │ancestral=deme1   │derived population to the ancestral    │
║    │       │            │                  │'deme1' population. Also set 'deme2'   │
║    │       │            │                  │to inactive, and all migration rates   │
║    │       │            │                  │to and from the derived population to  │
║    │       │            │                  │zero.                                  │
║    └───────────────────────────────────────────────────────────────────────────────┘
╠════════════════════════════════════╗
║ Epoch[2]: [3e+04, inf) generations ║
╠════════════════════════════════════╝
╟    Populations (total=3 active=1)
║    ┌───────────────────────────────────────────┐
║    │       │     start│       end│growth_rate  │
║    ├───────────────────────────────────────────┤
║    │  deme1│   10000.0│   10000.0│ 0           │
║    └───────────────────────────────────────────┘

```
